apps/zotonic_mod_wires/priv/lib/js/apps/zotonic-wired.js

```diff
@@ -23,621 +23,695 @@ Based on nitrogen.js which is copyright 2008-2009 Rusty Klophaus
-            });
-
-    // Start bridge to opener if opened from a window with a cotonic broker
-    if (window.opener && typeof window.opener.cotonic === 'object') {
-        cotonic.mqtt_bridge.newBridge('opener', {
-            client_id: '',
-            clean_start: true
-        });
-    }
-
-    setInterval(function() { z_transport_queue_check(); }, 500);
-}
-
-
-/* Non modal dialogs
----------------------------------------------------------- */
-
-function z_dialog_open(options)
-{
-    $.dialogAdd(options);
-}
-
-function z_dialog_close(options)
-{
-    if (typeof options === "number") {
-        $.dialogClose({ level: options });
-    } else if (options === "top") {
-        $.dialogClose({ level: "top" });
-    } else {
-        $.dialogClose(options);
-    }
-}
-
-function z_dialog_confirm(options)
-{
-    var html,
-        backdrop,
-        is_danger,
-        btn_class;
-
-    if (typeof options.backdrop == 'undefined') {
-        backdrop = options.backdrop
-    } else {
-        backdrop = true;
-    }
-    is_danger = options.is_danger || false;
-    btn_class = is_danger ? "btn-danger" : "btn-primary";
-    html = '<div class="confirm">' + options.text + '</div>'
-         + '<div class="modal-footer">'
-         + '<button class="btn btn-default z-dialog-cancel-button">'
-         + (options.cancel||z_translate('Cancel'))
-         + '</button>'
-         + '<button class="btn '+btn_class+' z-dialog-ok-button">'
-         + (options.ok||z_translate('OK'))
-         + '</button>'
-         + '</div>';
-    $.dialogAdd({
-        title: (options.title||z_translate('Confirm')),
-        text: html,
-        width: (options.width),
-        backdrop: backdrop,
-        level: options.level ?? "top"
+window.addEventListener(
+  "load",
+  function () {
+    const elts = document.querySelectorAll("link[media='none']");
+    elts.forEach(function (e) {
+      if (e.hasAttribute("media-onload")) {
+        e.setAttribute("media", e.getAttribute("media-onload"));
+        e.removeAttribute("media-onload");
+      }
     });
-    $(".z-dialog-cancel-button").click(function() {
-        z_dialog_close();
-        if (options.on_cancel) options.on_cancel();
-    });
-    $(".z-dialog-ok-button").click(function() {
-        z_dialog_close();
-        if (options.on_confirm) options.on_confirm();
-    });
-}
-
-function z_dialog_alert(options)
-{
-    var html,
-        backdrop;
-
-    if (typeof options.backdrop == 'undefined') {
-        backdrop = options.backdrop
+  },
+  false,
+);
+
+function zotonic_eval(script) {
+  if (script) {
+    if (z_script_nonce) {
+      // nonce - assume CSP headers set
+      const s = document.createElement("script");
+      const head = document.getElementsByTagName("head")[0];
+      const prev = document.getElementById("zotonic-eval-" + z_script_nonce);
+      if (prev) {
+        prev.remove();
+      }
+
+      s.type = "text/javascript";
+      s.id = "zotonic-eval-" + z_script_nonce;
```

Comment:
This provides the nonce value to javascript

Reply:
That doesn't really give new capabilities. Once an attacker has script access it can simply get the nonce by getting it from a script tag.

```javascript
document.getElementsByTagName("script")[0].nonce
```

Reply:
Exactly what @mmzeeman writes - if you can run JS (and the DOM), then you have access to the nonce by inspecting script/style tags.

So, the nonce is just a lock on the proverbial Javascript door, if you are inside then you have full access to everything (and all keys).

Reply:
Leaving this as "unresolved" so that the discussion is clear.

Reply:
Accessed like that you should find an empty value returned. See https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/nonce#accessing_nonces_and_nonce_hiding

Putting the nonce in an id however subverts this protection. 

I misspoke slightly in that the nonce can be had from javascript, albeit not like that, (see above link, and this is true in any case if you are programmatically generating the script tag and adding the nonce to it) but putting in a script id allows *other* things to potentially access it which is why it, as a tag attribute, is hidden.

Reply:
Ah, that is what you meant.   I will change that and make the tag id a bit less "unique".

Reply:
@CyBeRoni replaced it with one of the unique ids generated by Zotonic.